### DISCUSSION TITLE
Dim volume in multiplayer lobbies while waiting for game to start

### DISCUSSION THREAD

**Discussion by OliBomby** [ppy/osu#13173] [Ideas]
**Describe the new feature:** 
Optional feature to turn down volume of the music during multiplayer lobbies while not in-game. The music should return to normal level when playing or when leaving the lobby.

This feature would be usefull to me because I often want to listen to other stuff or talk (without tabbing out of osu) while I'm waiting for the multiplayer lobby to start a new map, so I turn down the volume manually. Then when the map actually starts I forget to turn it back to normal volume and I mess up my gameplay.

**Proposal designs of the feature:** 
This option can be added in the volume section of the options similar to the "Master (window inactive)" slider, but as this could add too many sliders in the future I propose changing the "Master (window inactive)" to something generic like "Master (dimmed)" and adding additional radio buttons for "Dim volume when window inactive" and "Dim volume when in multiplayer lobby" which control when to dim the volume.

If this design gets the greenlight then I might implement it myself.

**Comment by bdach** [comment]
I'm pretty sure the volume should be dimmed if the window is not in focus. If that is not happening for you, then that might be a framework bug - do you mind stating your platform and whether you can observe the volume dimming when changing the focused window?

**Comment by bdach** [comment]
On re-reading I might have misunderstood by what you mean by "in game". You mean "in the room but not actively playing", right?

I'll rename the issue to reflect that better. Not sure whether it should be a separate slider, probably just a checkbox...?

**Comment by OliBomby** [comment]
The volume does get dimmed when the window is not in focus. There are no issues with that. 
In this issue I ask for volume dimming while the window **is** in focus. That's why I added "(without tabbing out of osu)" in the description. I think you misunderstood my wording a bit.

Volume should be dimmed when you are in the multiplayer Room screen with the osu! window in focus.

**Comment by OliBomby** [comment]
> Not sure whether it should be a separate slider, probably just a checkbox...?

I had the same thought, but it's also nice to be able to control how much dimming happens, so I proposed changing the "Master (window inactive)" to something generic like "Master (dimmed)" and adding additional radio buttons for "Dim volume when window inactive" and "Dim volume when in multiplayer room" which control when to dim the volume.

That way you can control **when** and **how much** dimming should happen and it leaves room to add more areas to control the volume dimming.

**Comment by bdach** [comment]
To be honest, from our side, the less setting menu items the better, really. The settings menu is super cluttered as is anyway; I'd prefer it to apply automatically (maybe after the player has readied up) and universally (with no setting to control it) over a setting, and a single setting (can be a combo box) over two radio buttons.

**Comment by OliBomby** [comment]
I see why you want less stuff in the settings menu, but this volume dimming while waiting for a game to start is preference, so there still should be some control over it.

The problem with automatic dimming when readied up is people who want to listen to the music have to deal with the dimming and people who want the dimming have to wait for the next person to pick a song. 

Maybe the problem isn't that big. I forgot if you can ready up at any time in this version of multiplayer, but at least in osu! stable you can't ready up when the host is picking a song.
For people who want to listen to the music, they can just set the dimmed volume to 100% with the already existing slider.

I think this is an approach that can work, but it's also introducing these convoluted workarounds for control of the dimming which might be more confusing than 2 extra options in the settings menu. 
I personally think the settings menu isn't very cluttered if you use the search function effectively and I like having options for everything.

**Comment by peppy** [comment]
I'd rename the "inactive" volume setting to "dimmed" and then just have checkboxes or a dropdown for when the dimming should happen. One issue is that we only want to dim the music and hitsounds, but not interface sounds, in the case of multiplayer. We may need to make some changes to how the dimming is implemented to change this.

**Comment by smoogipoo** [comment]
I really don't understand this / don't agree with this. @OliBomby just to confirm, you _dim_ the track to listen to something else, instead of muting altogether? Do you also dim the other application when the match starts?

Imo dimming like this should be a user action, not something the game does for you.

**Comment by peppy** [comment]
The proposal is that this should be done by the game in the lobby, because it can happen that you are waiting for the lobby to start for 10-20 minutes while listening to the same loop of a 1 minute song the whole time.

**Comment by smoogipoo** [comment]
I understand the proposal, I just don't understand the reasoning. If you're waiting 10-20 minutes for a lobby to start, it takes 1 second to open the music controller and press pause. More UX like this just gets complicated/messy/unintuitive/leads to cognitive overload.
Because, for example, what if you don't want this in one particular case? Should an option then be added to quick-disable this setting because the user actually wants to listen to the track at full volume? Do we then make it full volume only for the first loop but then dim it for the second-onwards? It basically just becomes a whole tree of UX implementation for something that's effectively one click away.

**Comment by peppy** [comment]
I agree it's pretty bad UX. Leaving this in a proposal state for that reason (wouldn't consider unless it gains traction).

**Comment by smoogipoo** [comment]
I only brought it up because there's a PR already active, with the author asking for comments/a path forward/looking to make the changes to multiplayer itself (hence leading to another PR), and us disagreeing with the path taken.

Will close PR and redirect back here for further discussion.

**Comment by OliBomby** [comment]
The reasoning behind my proposal is that when I play osu! multi with friends, I like to talk over discord while waiting for the game to start and when we start playing we don't talk anymore. It's specific to the multiplayer waiting screen. The music pause button is hidden behind the "now playing" menu, so it's 2 clicks every time you get sent to the multiplayer waiting screen and I rather still have dampened music volume for bg music, so dampening music is better than pausing it completely. If the game dampens audio automatically that would be much nicer, because I dont have to manually correct the audio myself every single time.

About the UX concern, it doesn't have to be that complicated. If it's just one opt-in option, users can enable it for their convenience and disable it when they want to blast their ears with music again. 
Or we could have no option at all: Just always dampen music in the multiplayer waiting screen. If someone for whatever reason wants to listen to the music full volume, they can turn up the volume with Alt+scroll and the game will un-dampen the music. I think this design makes sense because the first thing someone will do if they want to turn up the volume, is to turn up the volume.

**Comment by 20Attila03** [comment]
I understand that it could be a bit complicated but I agree with the proposal here.
If someone wants to listen to that one music, they'll turn of the option and or turn up the volume.
As it is an opt-in / opt-out option with one button enabling it, I don't feel like this would be a problem. Also there is no need for too much control, in the case when you are waiting in a multiplayer lobby the dampening effect should only apply to the music as it is the music that may be frustrating for some.
I usually pause the music too as I want to talk to friends and also I don't want to hear a music that someone else puts in before playing as I would love to experience the music during the actual game.

**Reply by smoogipoo** [reply]
> If someone wants to listen to that one music, they'll turn of the option and or turn up the volume

What would "turn up the volume" achieve? We're back to where we started then. The only option is to turn off the option.

**Reply by 20Attila03** [reply]
Because the game is active, player's first action would be to try and scroll upwards, raising the volume. It can be an option to undo the effect.

As it is a new possible feature that we are talking about, you may need to implement a new feature to support that. How it would effect the volume settings or is it temporary and after the game starts it goes back to what the settings say.
We are talking about possible solutions for the few cases when a player would like to listen to that music at full volume and still have the option turned on.

**Comment by smoogipoo** [comment]
I completely understand the reasoning here - I also have played in tournaments where hearing my teammates was of upmost importance and had to pause music for this. But I would have _preferred_ to pause anyway because people have different audio levels/different tones/etc.
I strongly feel that if such an option were added, people would pause the music anyway. If so, the existence of the only causes complexity and UX overload with use by a very small percentage of users.

**Reply by 20Attila03** [reply]
As you may choose the volume for the cases, you can actually mute the music in the lobby too.
Also there are people who may prefer still hearing the music but not as loud as they would during gameplay, so they can talk on their preferred voice chatting platform during they're in the lobby.

**Comment by LittleEndu** [comment]
If the current PR were to move the volume slider to be only visible in multiplayer rooms, would that be acceptable way to avoid settings clutter?

**Reply by LittleEndu** [reply]
If it wasn't clear I meant move it out of settings panel entirely.

**Reply by OliBomby** [reply]
Does that mean there should be a settings panel for every individual screen separate from the main settings overlay? That will prevent cluttering the settings with volume sliders for many different screens.

**Reply by bdach** [reply]
I would be quite loath to see separate settings panels for separate screens. That's opening the door to another horde of incredibly specific settings that will barely get used.

**Comment by smoogipoo** [comment]
One way I could maybe see this happening is for a separate secondary volume control to appear while you're in a multiplayer room, and only while you're in a multiplayer room. Scrolling up and down would dim the master (global) volume, and you'd have to specifically hover the meter to change it (as you do for music/effect volume).

For implementation, it would be specific to a config manager managed by multiplayer, and the management of it would be handled by multiplayer itself. I would not want to see this inside `OsuConfigManager` and it wouldn't be enable per-screen settings as suggested by other users.
I imagine it similar to the way `DrawableRuleset` makes its config tracked globally, except instead it would be multiplayer adding a volume control:
https://github.com/ppy/osu/blob/be0605450b22f2d5cee868dd25c90095cc40de59/osu.Game/Rulesets/UI/DrawableRuleset.cs#L141-L151

**Reply by peppy** [reply]
i do not want more volume controls nested in a local interface.

this whole discussion feels way too drawn out for what it's attempting to fix (a minor annoyance). feels like a waste of time when i disagree with most of the presented solutions and PRs.

we will dim the volume automatically and not give customisation over it. i'll make an implementation of this in the future.

**Reply by smoogipoo** [reply]
I don't mean nested, and I don't mean a local interface.

Dim automatically? What's the point? What people want is to dim to the level where they can hear their friends over discord. It's impossible for us to figure this value unless providing customisation or dimming to a level way below what a lot of people will find acceptable. And what if I don't want it to dim, what if I'm not talking with friends at the same time?

These are the issues that we're trying to figure out here. Please engage in the discussion _first_ before making a PR, otherwise it's just going to waste time.

**Reply by peppy** [reply]
These are the things I would want to see happen, as individual changes:

- Volume control is added to toolbar. This gives users more visible control over game volume, and a single click to mute it if required. This requires experimentation and (minor) design work.
- Music volume is automatically dimmed while in the lobby and results screen by default, for all users, to some extent. Or potentially replaced by a new default implementation (we are looking at having more music loops as part of the game, rather than playing the song on loop). Or applies a high pass filter, once we have that support merged. This depends on further experimentation.

Main take away is that there should not be any new configuration settings specific to the lobby screen. This should not be a consideration made only for this one screen, not should it be attempted to be introduced as a new setting for the time being because it's clear this is going to add complexity that should be avoided.

If after the above there are still users wanting more control, maybe we can use the existing "play/pause" preview button which is on the beatmap listing cards, showing it on the current playing beatmap card in the lobby and allowing it to control whether the beatmap is playing. So, the control the user has would not be "dim" but "stop or start". And it wouldn't save anywhere.

But again, this will only affect the lobby. Not the results screen where the "can't hear people over the music" concern is still there.
